pkg/services/grpcserver/auth.go

```diff
@@ -0,0 +1,112 @@
+package grpcserver
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/grafana/grafana/pkg/cmd/grafana-cli/logger"
+	apikeygenprefix "github.com/grafana/grafana/pkg/components/apikeygenprefixed"
+	"github.com/grafana/grafana/pkg/infra/log"
+	"github.com/grafana/grafana/pkg/models"
+	"github.com/grafana/grafana/pkg/services/sqlstore"
+
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/metadata"
+	"google.golang.org/grpc/status"
+)
+
+// Authenticator can authenticate GRPC requests.
+type Authenticator struct {
+	logger   log.Logger
+	SQLStore sqlstore.Store
+}
+
+func NewAuthenticator(sqlStore sqlstore.Store) *Authenticator {
+	return &Authenticator{
+		logger:   log.New("grpc-server-authenticator"),
+		SQLStore: sqlStore,
+	}
+}
+
+// Authenticate checks that a token exists and is valid. It stores the user
+// metadata in the returned context and removes the token from the context.
+func (a *Authenticator) Authenticate(ctx context.Context) (context.Context, error) {
+	return a.tokenAuth(ctx)
+}
+
+const tokenPrefix = "Bearer "
+
+func (a *Authenticator) tokenAuth(ctx context.Context) (context.Context, error) {
+	auth, err := extractAuthorization(ctx)
+	if err != nil {
+		return ctx, err
+	}
+
+	if !strings.HasPrefix(auth, tokenPrefix) {
+		return ctx, status.Error(codes.Unauthenticated, `missing "Bearer " prefix in "authorization" value`)
+	}
+
+	token := strings.TrimPrefix(auth, tokenPrefix)
+	if token == "" {
+		return ctx, status.Error(codes.Unauthenticated, "token required")
+	}
+
+	newCtx := purgeHeader(ctx, "authorization")
+
+	err = a.validateToken(ctx, token)
+	if err != nil {
+		logger.Warn("request with invalid token", "error", err, "token", token)
+		return ctx, status.Error(codes.Unauthenticated, "invalid token")
+	}
+	return newCtx, nil
+}
+
+func (a *Authenticator) validateToken(ctx context.Context, keyString string) error {
```

Comment:
@FZambia @ryantxu i added support for service account tokens based on this message:
<img width="344" alt="image" src="https://user-images.githubusercontent.com/360020/176317319-db7b68e0-fb9d-4d96-9758-e398b66c31c5.png">

service account tokens can be set to never expire:
<img width="440" alt="image" src="https://user-images.githubusercontent.com/360020/176317535-305fcd13-670e-45cd-a7c5-58b4c96f3de0.png">

it looks like API keys still exist, but a TTL is required:
<img width="722" alt="image" src="https://user-images.githubusercontent.com/360020/176317473-9c88705b-96f6-480d-9115-fc3873747ed3.png">